core/src/main/scala/org/apache/spark/scheduler/BlacklistStrategy.scala

```diff
@@ -0,0 +1,156 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.scheduler
+
+import scala.collection.mutable
+
+import org.apache.spark.Logging
+import org.apache.spark.SparkConf
+import org.apache.spark.util.SystemClock
+
+/**
+ * The interface to determine executor blacklist and node blacklist.
+ */
+trait BlacklistStrategy {
+  val expireTimeInMillisecond: Long
+
+  // TaskId is optional here because the BlacklistStrategy could be unrelated to TaskId
```

Comment:
I think I'm still missing something.  Whats the point of cases 3 & 4?  Is there a situation in which the caller wouldn't have access to a task id, and so couldn't provide it?  The only place I see it called, it will always get called with `Some(taskId)`.  And if there is, would it actually make more sense as a different method?

Otherwise this seems like unusual interface design.  The caller shouldn't care what the actual implementation does -- it just passes in a the task id, b/c the implementation *might* need it.  Implementations are free to ignore it.  In fact, this even means that in situation (2), you shouldn't be logging a warning either, its expected that the caller will give you a task id even if you don't need it.

I do agree that we should `Option` rather than `null` if we do need that flexibility, I'm just not seeing why its necessary.